docs/src/references/agent/operation-workflow.md

```diff
@@ -328,6 +332,56 @@ Note that:
 - No `on_exit` nor `on_kill` status can be provided, as the script is not monitored.
 - If the script cannot be launched, the workflow will be moved to the final `"failed"` state.
 
+### Sub-Command Execution
+
+An operation workflow can trigger a command defined by another workflow.
+
+```toml
+["<state-name>"]
+command = "<operation-name>"
+```
+
+Operation specific arguments can be provided to the sub-command instance,
+and data extracted from the calling command status topic and message payload can be injected into these arguments.
+
+- For instance, `cmd_arg.url = "http://127.0.0.1:8000/tedge/file-transfer/foo.txt"`
+  adds an `url` property to the init message of the sub-command.
+- Similarly, `cmd_arg.url = "${.payload.config_update.url}"`
+  adds an `url` property to the init message of the sub-command, but extracting this url from the calling command state.
+- Several such arguments can be added to the sub-command initial state.
+  - All these individual arguments are combined into a JSON object value published on the sub-command MQTT topic.
+  - The sub-command initial state can be provided as a whole with `cmd_arg = ${.payload.path.to.sub-command.init-status}`
+  - The calling command excerpts injected into this initial state are defined using the same path conventions as for script handlers.
+  - If a `cmd_arg.status` value is provided, this value will be ignored and replaced by `"init"`
+    as this is the only valid value for the initial state of a command.
+  - If no value at all is provided for `cmd_arg`, then the default value is `{"status" = "init"}`.
```

Comment:
Providing the arguments properties in the current state works for trivial payloads. I see a few different options to handle no-trival arg conversions when calling a sub-command:

**Option 1:** Any prior conversion must be done in the previous state

```toml
[prepare_software]
command = "/usr/bin/format_software_list.sh"
success = "update_software"

[update_software]
command = "/usr/bin/start_software_update.sh"
```

**Option 2:** Allow users to provide a command which returns the requested sub-command with arguments.

Users can provide a command which will return the data to be used to call the sub-command. This would at least remove the need to have a "prepare_xxx" command before the state which actually does the request (though it would make it a little harder to track).

```toml
[update_software]
command = "/usr/bin/start_software_update.sh"
```

And the command would return some well formed command request structure (similar to the begin/end blocks for the state modifications)

```sh
#!/bin/sh
echo --- tedge command begin ---
echo command=software_update
echo args.value1=$(som_external_func)
echo args.value2.other={"foo":"bar"}
echo --- tedge command end ---
```

**Option 3:** Support more complex template language that can be used to build the command args.

Reply:
> Providing the arguments properties in the current state works for trivial payloads. I see a few different options to handle no-trival arg conversions when calling a sub-command:

It seems that I failed to convey the idea. In `cmd_arg.x = "${.payload.x}"` the right-end-side expression can be any JSON value, not only a scalar but also an array or an object or a combination of these. 
 
> **Option 1:** Any prior conversion must be done in the previous state
> 
> ```toml
> [prepare_software]
> command = "/usr/bin/format_software_list.sh"
> success = "update_software"
> 
> [update_software]
> command = "/usr/bin/start_software_update.sh"
> ```

This is exactly the plan, assigning to `cmd_arg` pre-computed values, stored in the main operation state.

```
[update_software]
command = "update_software"
cmd_args = "${.payload.software_list}"
```

> **Option 2:** Allow users to provide a command which returns the requested sub-command with arguments.
> 
> Users can provide a command which will return the data to be used to call the sub-command. This would at least remove the need to have a "prepare_xxx" command before the state which actually does the request (though it would make it a little harder to track).
> 
> ```toml
> [update_software]
> command = "/usr/bin/start_software_update.sh"
> ```
> 
> And the command would return some well formed command request structure (similar to the begin/end blocks for the state modifications)
> 
> ```shell
> #!/bin/sh
> echo --- tedge command begin ---
> echo command=software_update
> echo args.value1=$(som_external_func)
> echo args.value2.other={"foo":"bar"}
> echo --- tedge command end ---
> ```

Why not. This will avoid to clutter the state of the calling operation with arguments for the invoked operation. However, I will keep the same conventions as for scripts (i.e. tedge begin / end) the name of the command being defined in the main operation workflow.

```
command = "software_update"
cmd_args_from = "/script/extracting/software_update/init_state/from/current_state.sh"
``` 
 
> **Option 3:** Support more complex template language that can be used to build the command args.

We need first more examples. If it happens that most of the `/script/extracting/software_update/init_state/from/current_state.sh` are using `jq`, then we can make `jq` first citizen.

Reply:
> Option 2: Allow users to provide a command which returns the requested sub-command with arguments.

> Users can provide a command which will return the data to be used to call the sub-command. This would at least remove the need to have a "prepare_xxx" command before the state which actually does the request (though it would make it a little harder to track).

This has been implemented  by  https://github.com/thin-edge/thin-edge.io/pull/2792/commits/b3dae720e140fc8938766642bd1ee9c36f2965c5